### PR TITLE
Do not shutdown the async initializer executor

### PR DESCRIPTION
**Pull request by fsamuel-bs**
**Goals (and why)**: Do not shutdown `AsyncInitializer` executor as there might have been a `cancelInitialization` call to the initializing object, and we need to be able to process its handler.

**Implementation Description (bullets)**: -

**Concerns (what feedback would you like?)**: -

**Where should we start reviewing?**: -

**Priority (whenever / two weeks / yesterday)**: -

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2547)

### CODE REVIEW

File: docs/source/release_notes/release-notes.rst [hunk 1]
```diff
@@ -63,11 +63,6 @@ develop
            (`Pull Request <https://github.com/palantir/atlasdb/pull/2518>`__)
```
Comment (gsheasby) [inline docs/source/release_notes/release-notes.rst]: NB this PR ref is wrong - should be 2520.

File: atlasdb-commons/src/main/java/com/palantir/async/initializer/AsyncInitializer.java [hunk 2]
```diff
@@ -71,7 +75,6 @@ private void tryInitializationLoop() {
                     SafeArg.of("className", getInitializingClassName()),
                     SafeArg.of("numberOfAttempts", numberOfInitializationAttempts),
                     SafeArg.of("initializationDuration", System.currentTimeMillis() - initializationStartTime));
-            singleThreadedExecutor.shutdown();
```
Comment (gsheasby) [inline atlasdb-commons/src/main/java/com/palantir/async/initializer/AsyncInitializer.java]: can we ever shut this down?

Comment (fsamuel-bs) [reply atlasdb-commons/src/main/java/com/palantir/async/initializer/AsyncInitializer.java]: Maybe you could close all of them when `SerializableTransactionManager` is initialized, as you don't expect any of them to be closed in the middle of a initialization anymore... but I'm not sure. @nziebart for SA.

Comment (nziebart) [reply atlasdb-commons/src/main/java/com/palantir/async/initializer/AsyncInitializer.java]: would be nice to have a test for this

Comment (nziebart) [reply atlasdb-commons/src/main/java/com/palantir/async/initializer/AsyncInitializer.java]: also - wonder if we should add a try/catch around `cleanupOnInitFailure`. If something throws in there, currently we would just die and never initialize

**Comment by nziebart** [comment]
have we done tests around how long it takes to initialize if the kvs is down? given that we sleep for 10 sec between iterations, and our dependency graph is 4(?) levels deep, seems like we could wait a good while unnecessarily.

perhaps each service could control its sleep time. 10 sec for cassandra is probably fine, but for the others, they are basically just waiting on the KVS, and could try much more frequently

**Comment by nziebart** [comment]
Here's a crazy idea - why even have the executors? We could just try to initialize every time you call a method, with rate limiting. And then we can have a single background thread on the transaction manager (topmost object in the graph) that tries to initialize it

**Comment by gsheasby** [comment]
The "crazy idea" could work, but I'd push for having it in a separate PR - it sounds like it would make initialisation more quicker, but even the current state of the new feature (perhaps up to 40 seconds to initialise) is better than the base state (never initialising).

File: atlasdb-commons/src/main/java/com/palantir/async/initializer/AsyncInitializer.java (hunk 2 above)
Comment (gsheasby) [reply atlasdb-commons/src/main/java/com/palantir/async/initializer/AsyncInitializer.java]: perhaps `try {cleanupOnInitFailure} finally {scheduleInitialization()}`, so we at least always schedule?

**Comment by gsheasby** [comment]
// handing off to NZiebart as he did have some concerns. I think it's fine to merge as-is, though (retriggered the build as it seemed to flake)

**Comment by fsamuel-bs** [comment]
While the "crazy idea" would work for classes such as the CKVS, it wouldn't work for classes that assume that they are going to eventually get initialized, and don't have any public API, like the TimestampTracker.

We could keep the executor approach just for these kinds of classes, but then we would need to maintain two implementations. So I wouldn't implement it at this time.

**Comment by nziebart** [comment]
It wouldn't need to be two implementations, it's just an option to have an executor polling you.

I'm fine with not doing this for now though, but would push for at least something to mitigate the total time taken, like my suggestion to just make the polling interval configurable.

**Comment by fsamuel-bs** [comment]
If we want to improve the time it takes for us to initialize (which is not the goal of this PR), I think we should:

1. have a metric that tracks this — probably a benchmark, since all our options seem prone to develop error in the future, and this would be a metric that would be nice to have (we just log this, currently);
2. since we now have a `isInitialized()` method in each object being initialized, we could also check that your dependencies are initialized as the first thing on `tryInitialize()`, and lower the pooling interval (by also making it configurable, as per your suggestion).

Anyhow, I think we should track these possible solutions as an issue and merge this PR.

**Comment by fsamuel-bs** [comment]
Filed #2600.

**Pull request merged**